newlib/libc/machine/riscv/memcpy.S

```diff
@@ -0,0 +1,295 @@
+/* Copyright (c) 2019  SiFive Inc. All rights reserved.
+
+   This copyrighted material is made available to anyone wishing to use,
+   modify, copy, or redistribute it subject to the terms and conditions
+   of the FreeBSD License.   This program is distributed in the hope that
+   it will be useful, but WITHOUT ANY WARRANTY expressed or implied,
+   including the implied warranties of MERCHANTABILITY or FITNESS FOR
+   A PARTICULAR PURPOSE.  A copy of this license is available at
+   http://www.opensource.org/licenses.
+*/
+
+#include <sys/asm.h>
+
+.text
+.global memcpy
+.type	memcpy, @function
+memcpy:
+#if defined(PREFER_SIZE_OVER_SPEED) || defined(__OPTIMIZE_SIZE__)
+  mv a3, a0
+  beqz a2, 2f
+
+1:
+  lbu a4, 0(a1)
+  sb a4, 0(a3)
+  add   a2, a2, -1
+  add   a3, a3, 1
+  add   a1, a1, 1
+  bnez a2, 1b
+
+2:
+  ret
+
+#else
+    mv a3, a0
+    li    a4, SZREG
+    addi  sp, sp, -2 * SZREG
+#if __riscv_xlen == 32
+    sw    s0, 0(sp)
+    sw    s1, 4(sp)
+#else
+    sd    s0, 0(sp)
+    sd    s1, 8(sp)
+#endif
+    bleu  a2, a4, .Lcopy1
+
+
+#if defined(__riscv_misaligned_avoid) || defined(__riscv_misaligned_slow)
+    andi a4, a1, SZREG - 1
```

Comment:
i have a feeling these three instructions may be optimized. 
e.g. if you replace the andi, andi, sub sequence with an xor, andi SZREG - 1, i think this should work

Reply:
actually i think an xnor will be most efficient

Reply:
seeing now that you use a4 later, maybe this is not a huge benefit. but at least you can move it into the path that needs it, so for the path that branches to Lalign, there could be less instructions.